app/routes.py

```diff
@@ -1,2 +1,315 @@
+def update_task(task_id):
+    """
+    Performs the update action and returns a 200 response with the newly updated task
+
+    If any of the expected field is missing in the post request body, it returns
+    a 400 response indicating invalid data
+
+    Returns a 404 with no response body in case the task is not found
+    """
+    task = Task.query.get_or_404(task_id)
+
+    request_body = request.get_json()
+
+    if invalid_post_request_body(request_body):
+        return make_response({"details": "Invalid data"}, 400)
+
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+    task.completed_at = request_body["completed_at"]
+    db.session.commit()
+    return make_response(jsonify(task=task.as_dict()), 200)
+
+
+@tasks_bp.route("/<int:task_id>/mark_complete", methods=["PATCH"])
+def mark_task_complete(task_id):
+    """
+    Changes the task.completed_at value to todays date and returns a 200 response
+    with the updated task and calls the send_task_notification funtion
+    Returns a 404 with no response body in case the task is not found
+    """
+    task = Task.query.get_or_404(task_id)
+
+    task.completed_at = datetime.now(timezone.utc)
+    db.session.commit()
+
+    send_slack_task_notification(task)
+    return make_response(jsonify(task=task.as_dict()), 200)
+
+
+def send_slack_task_notification(task):
+    """
+    Sends a request to a slack bot to post the
+    "Someone just completed the task <TASK_TITLE>" to the task-notifications channel
+    in the configured slack workspace
+    """
+    SLACK_BOT_TOKEN = os.environ.get("SLACK_BOT_TOKEN")
+    text = f"Someone just completed the task {task.title}"
+    url = f"https://slack.com/api/chat.postMessage?channel=task-notifications&text={text}"
+
+    payload = ""
+
+    headers = {
+        'Authorization': f'Bearer {SLACK_BOT_TOKEN}'
+    }
+
+    return requests.request("POST", url, headers=headers, data=payload)
+
+
+@tasks_bp.route("/<int:task_id>/mark_incomplete", methods=["PATCH"])
+def mark_task_incomplete(task_id):
+    """
+    Changes the task.completed_at value to None and returns a 200 response
+    with the updated task
+    Returns a 404 with no response body in case the task is not found
+    """
+    task = Task.query.get_or_404(task_id)
+
+    task.completed_at = None
+    db.session.commit()
+    return make_response(jsonify(task=task.as_dict()), 200)
+
+
+@goals_bp.route("", methods=["GET"])
+def list_all_goals():
+    """
+    Returns a 200 response with a list of all goals currently
+    in the database as the response body
+    Optional query arguments:
+     * sort = asc or desc
+    """
+
+    query_param_value = request.args.get("sort")
+    if query_param_value == "desc":
+        goals = Goal.query.order_by(Goal.title.desc()).all()
+    elif query_param_value == "asc":
+        goals = Goal.query.order_by(Goal.title).all()
+    else:
+        goals = Goal.query.all()
```

Comment:
hmm this might be a good chunk of code we could turn into a helper method inside Goal model